modules/github_integration/app/workers/cron/check_deploy_status_job.rb

```diff
@@ -0,0 +1,183 @@
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Cron
+  class CheckDeployStatusJob < ApplicationJob
+    include OpenProject::GithubIntegration::NotificationHandler::Helper
+
+    priority_number :low
+
+    def perform
+      deploy_targets.find_each do |deploy_target|
+        sha = openproject_core_sha deploy_target.host, deploy_target.api_key
+
+        if sha.present?
+          pull_requests.find_each do |pull_request|
+            check_deploy_status deploy_target, pull_request, sha
+          end
+        else
+          OpenProject.logger.error "Failed to retrieve core SHA for deploy target #{deploy_target.host}"
+        end
+      end
+    end
+
+    def deploy_targets
+      DeployTarget.all
+    end
+
+    def pull_requests
+      GithubPullRequest
+        .closed
+        .where.not(merge_commit_sha: nil)
+    end
+
+    def check_deploy_status(deploy_target, pull_request, core_sha)
+      status_check = deploy_status_check deploy_target, pull_request
+
+      # we already checked this PR against the given core SHA, so no need to check again
+      return if status_check.core_sha == core_sha
+
+      # if the commit is contained, it has been deployed
+      if commit_contains? core_sha, pull_request.merge_commit_sha
+        update_deploy_status pull_request, deploy_target
+      else
+        status_check.update core_sha: # remember last checked SHA to not check twice
+      end
+    end
+
+    ##
+    # Marks the given PR as deployed and removes the last status check
+    # as it won't be needed anymore. This is because only closed (not yet deployed)
+    # PRs are ever checked for their deployment status.
+    def update_deploy_status(pull_request, deploy_target)
+      host = deploy_target.host
+
+      ActiveRecord::Base.transaction do
+        delete_status_check pull_request, deploy_target
+        pull_request.update! state: "deployed"
+
+        comment_on_referenced_work_packages(
+          pull_request.work_packages,
+          comment_user,
+          "[#{pull_request.repository}##{pull_request.number}](#{pull_request.github_html_url}) deployed to [#{host}](https://#{host})"
+        )
+      end
+    end
+
+    def delete_status_check(pull_request, deploy_target)
+      # we use `find` and delete it this way to also cover
+      # not-yet-persisted records
+      check = pull_request
+        .deploy_status_checks
+        .find { |c| c.deploy_target == deploy_target }
```

Comment:
Since we can have multiple `deploy_status_checks` with the same `deploy_target`, I would use `select` instead of `find` to remove all of them.
```suggestion
        .select { |c| c.deploy_target == deploy_target }
```